design/design-doc-resource-configurations.md

```diff
@@ -0,0 +1,838 @@
+may seem innocuous, it has undesirable properties:
+
+* It forces Crossplane's assumptions on how resources will be used; i.e. that a
+  Crossplane user would never want to create a GKE node pool without also
+  managing its control plane in Crossplane, or would never want to create a
+  virtual network rule for an Azure SQL server that Crossplane was unaware of.
+* It violates the [principle of least astonishment]. In the `S3Bucket` example
+  above it may be surprising for a Crossplane user to be provisioned an IAM user
+  when they only requested an S3 bucket.
+* The mapping from Crossplane managed resource to underlying API becomes less
+  obvious - it's not clear that virtual network rules are actually a distinct
+  API rather than an inherent part of a MySQL server. This makes it harder for
+  Crossplane users to fall back to the underlying provider's documentation when
+  Crossplane's is unclear.
+* Each managed resource reconciler becomes more complex, as it must reconcile
+  each managed resource with multiple external resources.
+
+Supporting complex managed resources would remove (or less generously, move) the
+need to add complexity to the Crossplane API and resource claim logic, but the
+cost would be a complex mental model of how Crossplane managed resources relate
+to cloud resources, and limitations placed on how Crossplane could be used.
+
+### Cross-claim References
+
+Cross-claim references would allow resource claims to reference other resource
+claims, similarly to how managed resources can reference other managed
+resources. Rather than a `MySQLInstance` claim binding to an `AggregateResource`
+of a `MySQLServer` and a `MySQLVirtualNetworkRule` each managed resource would
+be claimed separately - the `MySQLServer` by the `MySQLInstance` and the
+`MySQLVirtualNetworkRule` by a new claim: perhaps `FirewallRule`. The
+`FirewallRule` claim would state via a cross-claim reference that it applies to
+the `MySQLInstance` claim.
+
+This design shifts much of the burden of designing infrastructure from
+infrastructure operators to application operators. Instead of an application
+operator requesting a `MySQLInstance` and trusting that the infrastructure
+operator has published a resource class that will ensure it is securely
+configured, they must also request a `FirewallRule` be applied to that
+`MySQLInstance`. Furthermore, the "lowest common denominator" problem affects
+resource claims, in that they can only expose configuration fields that
+translate to _every_ managed resource that could satisfy the claim. A
+`FirewallRule` for example could only expose configuration fields that apply to
+`MySQLVirtualNetworkRule`, `SecurityGroup`, and any other firewall-like managed
+resource. This means application owners would simultaneously be saddled with the
+burden of designing appropriate infrastructure for their needs and a limited
+language in which to do so.
+
```

Comment:
We could possibly use validating webhooks for validations like `look at the apiVersion and Kind and send the object to api-server validation endpoint`

Reply:
Agreed. I had a few speculative ideas about this. Approximately in increasing order of difficulty to implement but also in increasing order of type safety:

* Use a plain old CRD to validate the scaffolding/general shape of a `ResourceConfigurationClass` (the fields that never change).
* Use a plain old CRD with `x-kubernetes-embedded-resource` and `x-kubernetes-preserve-unknown-fields`  to validate that the array of resources [quack like](https://youtu.be/kldVg63Utuw) a Crossplane managed resource (e.g. have a `classRef`, `claimRef`, `bindingPhase`, etc)?
* Use an admission control webhook to iterate over the array of resources and validate them against the CRD that defines their GVK.
* Use an admission control webhook to iterate over the array of patches to each resource and ensure the referenced "from claim" and "to resource" fields actually exist and are of compatible types (by inspecting their CRDs).

Reply:
I think it can be added as a future consideration linking https://github.com/crossplaneio/crossplane/issues/1152

Reply:
Yeah, I think ideally we'd land parts 1 and 2 (the parts that can be done with CRDs) on the initial pass of this design, then shore it up with a webhook in a subsequent release.